File: 36_graphs/36_5_reachability_queries.cpp
#include <iostream>
#include <vector>
#include <set>
#include <map>

void dfs(const std::vector<std::vector<int>>& graph, int cc_id, int node, std::set<int>& visited, std::map<int, int>& node_to_cc) {
  visited.insert(node);
  node_to_cc[node] = cc_id;
  for (size_t i = 0; i < graph[node].size(); i++) {
    int nei = graph[node][i];
    if (visited.find(nei) == visited.end()) {
      dfs(graph, cc_id, nei, visited, node_to_cc);
    }
  }
}

std::vector<bool> reachability_queries(const std::vector<std::vector<int>>& graph, const std::vector<std::pair<int, int>>& queries) {
  std::set<int> visited;
  std::map<int, int> node_to_cc;

  int cc_id = 0;
  for (int node = 0; node < graph.size(); node++) {
    if (visited.find(node) == visited.end()) {
      dfs(graph, cc_id, node, visited, node_to_cc);
      cc_id++;
    }
  }
  
  std::vector<bool> output;
  for (std::vector<std::pair<int, int>>::const_iterator it = queries.begin(); it != queries.end(); ++it) {
    output.push_back(node_to_cc[it->first] == node_to_cc[it->second]);
  }

  return output;
}

int main() {
    std::vector<std::vector<int> > graph;
    graph.push_back(std::vector<int>(1, 1));
    int arr1[] = {0, 2, 5, 4};
    graph.push_back(std::vector<int>(arr1, arr1 + 4));
    int arr2[] = {1, 4, 5};
    graph.push_back(std::vector<int>(arr2, arr2 + 3));
    graph.push_back(std::vector<int>());
    int arr3[] = {5, 2, 1};
    graph.push_back(std::vector<int>(arr3, arr3 + 3));
    int arr4[] = {1, 2, 4};
    graph.push_back(std::vector<int>(arr4, arr4 + 3));

    std::vector<std::pair<int, int> > queries;
    queries.push_back(std::make_pair(0, 4));
    queries.push_back(std::make_pair(0, 3));

    std::vector<bool> result = reachability_queries(graph, queries);

    std::cout << "[";
    for (size_t i = 0; i < result.size(); i++) {
        std::cout << (result[i] ? "true" : "false");
        if (i < result.size() - 1) std::cout << ", ";
    }
    std::cout << "]" << std::endl;

    return 0;
}

File: 36_graphs/36_5_reachability_queries.py
def reachability_queries(graph: list[list[int]], queries: list[int])->list[bool]:
  visited = set()
  node_to_cc = {}

  def dfs(graph: list[list[int]], cc_id: int, node: int):
    visited.add(node)
    node_to_cc[node] = cc_id
    for nei in graph[node]:
      if nei not in visited:
        dfs(graph, cc_id, nei)
  
  cc_id = 0
  for node in range(len(graph)):
    if node not in visited:
      dfs(graph, cc_id, node)
      cc_id += 1
  
  output = []
  for q1, q2 in queries:
    output.append(node_to_cc[q1] == node_to_cc[q2])

  return output

if __name__ == "__main__":
  graph = [
  [1],
  [0, 2, 5, 4],
  [1, 4, 5],
  [],
  [5, 2, 1],
  [1, 2, 4]
]
queries = [[0, 4], [0, 3]]
print(reachability_queries(graph, queries))